qiskit_nature/second_q/properties/electronic_density.py

```diff
@@ -103,35 +114,83 @@ def interpret(
-        n_spin = 2 * n_spatial
-
-        rdm1 = np.zeros((n_spin, n_spin), dtype=float)
-        rdm2 = np.zeros((n_spin, n_spin, n_spin, n_spin), dtype=float)
 
         rdm1_idx_regex = re.compile(r"RDM\((\d+), (\d+)\)")
         rdm2_idx_regex = re.compile(r"RDM\((\d+), (\d+), (\d+), (\d+)\)")
 
+        contains_rdm2 = "++--" in self.alpha.keys()
+
+        rdm1_a = np.zeros((n_spatial, n_spatial), dtype=float)
+        rdm1_b = np.zeros((n_spatial, n_spatial), dtype=float)
+
+        rdm2_aa: np.ndarray | None = None
+        rdm2_bb: np.ndarray | None = None
+        rdm2_ba: np.ndarray | None = None
+        if contains_rdm2:
+            rdm2_aa = np.zeros((n_spatial, n_spatial, n_spatial, n_spatial), dtype=float)
+            rdm2_bb = np.zeros((n_spatial, n_spatial, n_spatial, n_spatial), dtype=float)
+            rdm2_ba = np.zeros((n_spatial, n_spatial, n_spatial, n_spatial), dtype=float)
+
         for name, aux_value in result.aux_operators_evaluated[0].items():
             # immediately skip zero values
             if np.isclose(aux_value, 0.0):
                 continue
 
             match = rdm1_idx_regex.fullmatch(name)
             if match is not None:
-                mo_i, mo_j = (int(idx) for idx in match.groups())
-                rdm1[mo_i, mo_j] = aux_value.real
+                index = tuple(int(idx) for idx in match.groups())
+                bools = _boolean_index(index, n_spatial)
+
+                if all(bools):
+                    rdm1_a[index] = aux_value.real
+                elif not any(bools):
+                    rdm1_b[_shifted_index(index, n_spatial)] = aux_value.real
+
                 continue
 
-            match = rdm2_idx_regex.fullmatch(name)
-            if match is not None:
-                mo_i, mo_j, mo_k, mo_l = (int(idx) for idx in match.groups())
-                rdm2[mo_i, mo_j, mo_k, mo_l] = aux_value.real
+            if contains_rdm2:
+                match = rdm2_idx_regex.fullmatch(name)
+                if match is not None:
+                    index = tuple(int(idx) for idx in match.groups())
+                    bools = _boolean_index(index, n_spatial)
+
+                    if all(bools):
+                        rdm2_aa[index] = aux_value.real
+                    elif not any(bools):
+                        rdm2_bb[_shifted_index(index, n_spatial)] = aux_value.real
+                    elif bools[0] and bools[3] and not bools[1] and not bools[2]:
+                        rdm2_ba[_shifted_index(index, n_spatial)] = aux_value.real
 
         result.electronic_density = ElectronicDensity.from_raw_integrals(
-            rdm1[n_spatial:, n_spatial:],
-            rdm2[n_spatial:, n_spatial:, n_spatial:, n_spatial:],
-            rdm1[:n_spatial, :n_spatial],
-            rdm2[:n_spatial, :n_spatial, :n_spatial, :n_spatial],
-            rdm2[:n_spatial, n_spatial:, n_spatial:, :n_spatial],
+            rdm1_a,
+            rdm2_aa if contains_rdm2 else None,
```

Comment:
Is this check here needed? Above, lines 126 thru 132 rdm2_aa etc are initialized to None and then set based contains_rdm2 - so these vars should already be None or some valid value, based on contains_rdm2 logic above, that this ternary seems like it should not be needed again here.